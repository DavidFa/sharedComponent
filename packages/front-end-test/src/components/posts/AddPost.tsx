import styled from 'styled-components';
import useAddPost from './hooks/useAddPost';

const Wrapper = styled.div`
margin: 1rem auto;
width:400px;
border-radius: 4px;
border: 1px solid #ccc;
`;

const Form = styled.form`
display:flex;
flex-direction: column;
`;

const InputGroup = styled.div`
padding: 0.3rem;
display:inline-block;
`;

const Label = styled.label`
width:70px;
display:inline-block;
`;

const Input = styled.input`
width:300px;
`;

const Button = styled.button``;

const AddPost = () => {
    const { post, onTitleChangeHandler, onBodyhangeHandler, onSubmitHandler, onCancelHandler } = useAddPost();

    return (
        <Wrapper>
            <Form onSubmit={onSubmitHandler} role='form'>
                <InputGroup><h3>Edit Post</h3></InputGroup>
                <InputGroup><Label>Title</Label><Input data-testid='input-title' value={post.title} onChange={onTitleChangeHandler} /></InputGroup>
                <InputGroup><Label>Body</Label><Input data-testid='input-body' value={post.body} onChange={onBodyhangeHandler} /></InputGroup>
                <InputGroup><Button type="button" onClick={onCancelHandler}>Cancel</Button><Button>Save</Button></InputGroup>
            </Form>
        </Wrapper>
    )
}

export default AddPost;
