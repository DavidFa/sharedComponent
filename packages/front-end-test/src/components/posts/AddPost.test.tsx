import AddPost from './AddPost';
import { render, screen, fireEvent } from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';

describe('AddPost', () => {
    describe("Snapshot Test", () => {
        it("should render with given state", () => {
            const component = renderer.create(<AddPost />);
            expect(component.toJSON()).toMatchSnapshot();
        });

    });

    const mockUseDispatch = jest.fn();
    const setState = jest.fn();

    beforeEach(() => {
        jest.spyOn(ReactRedux, 'useDispatch').mockReturnValue(mockUseDispatch);

    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('', () => {
        it('should dispatch an action when click save button', () => {
            const useStateMock = (initState: { title: string, body: string }) => [initState = { title: "title", body: "body" }, setState];
            jest.spyOn(React, 'useState').mockImplementation(useStateMock);
            render(<AddPost />)
            // const jsdomAlert = window.alert;
            // window.alert = () => { };
            expect(screen.getByRole('form')).toBeInTheDocument();
            expect(screen.queryByText(/Add Post/i)).toBeInTheDocument();
            expect(screen.queryByText(/Title/i)).toBeInTheDocument();
            expect(screen.queryByText(/Body/i)).toBeInTheDocument();
            expect(screen.queryByText(/Body/i)).toBeInTheDocument();
            expect(screen.queryByTestId(/input-title/i)).toBeInTheDocument();
            expect(screen.queryByTestId(/input-body/i)).toBeInTheDocument();

            const cancel = screen.getByText(/Cancel/i);
            expect(cancel).toBeInTheDocument();

            fireEvent.click(cancel);
            expect(mockUseDispatch).toHaveBeenCalledTimes(1);

            const save = screen.getByText(/Save/i);
            expect(save).toBeInTheDocument();

            fireEvent.click(save);
            expect(mockUseDispatch).toHaveBeenCalledTimes(2);
            // window.alert = jsdomAlert;
        })
    })


    it('should not dispatch an action when click save button', () => {
        const useStateMock = (initState: { title: string, body: string }) => [initState = { title: "", body: "" }, setState];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        render(<AddPost />)
        const jsdomAlert = window.alert;
        window.alert = () => { };

        const save = screen.getByText(/Save/i);
        expect(save).toBeInTheDocument();

        fireEvent.click(save);
        expect(mockUseDispatch).toHaveBeenCalledTimes(0);
        window.alert = jsdomAlert;
    })
})