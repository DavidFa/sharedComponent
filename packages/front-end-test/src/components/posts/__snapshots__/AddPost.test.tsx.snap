// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPost Snapshot Test should render with given state 1`] = `
<div
  className="sc-bdfBQB eiRpoH"
>
  <form
    className="sc-gsTEea bSRuSa"
    onSubmit={[Function]}
    role="form"
  >
    <div
      className="sc-dlfnuX cmPjgH"
    >
      <h3>
        Add Post
      </h3>
    </div>
    <div
      className="sc-dlfnuX cmPjgH"
    >
      <label
        className="sc-hKgJUU bYFslP"
      >
        Title
      </label>
      <input
        className="sc-eCstlR hiWetG"
        data-testid="input-title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="sc-dlfnuX cmPjgH"
    >
      <label
        className="sc-hKgJUU bYFslP"
      >
        Body
      </label>
      <input
        className="sc-eCstlR hiWetG"
        data-testid="input-body"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="sc-dlfnuX cmPjgH"
    >
      <button
        className="sc-jSgvzq jZKsKI"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="sc-jSgvzq jZKsKI"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
